pkg/operator/controller/controller.go

```diff
@@ -111,127 +111,121 @@ type reconciler struct {
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dns 'cluster': %v", err)
+	}
+	infraConfig := &configv1.Infrastructure{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get infrastructure 'cluster': %v", err)
+	}
+	ingressConfig := &configv1.Ingress{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, ingressConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get ingress 'cluster': %v", err)
+	}
 
-		// For now, if the cluster configs are unavailable, defer reconciliation
-		// because weaving conditionals everywhere to deal with various nil states
-		// is too complicated. It doesn't seem too risky to rely on the invariant
-		// of the cluster config being available.
-		if dnsConfig != nil && infraConfig != nil && ingressConfig != nil {
-			// Ensure we have all the necessary scaffolding on which to place router instances.
-			if err := r.ensureRouterNamespace(); err != nil {
-				errs = append(errs, fmt.Errorf("failed to ensure router namespace: %v", err))
-			}
+	// Only consider the ingresscontroller safe to process if it has been defaulted
+	// and validated.
+	r.setIngressControllerDefaults(ingress, ingressConfig, infraConfig)
+	// TODO: Persist the defaults if there's a diff
+	if err := r.validateIngressController(ingress); err != nil {
+		// TODO: Set some sort of failing conditions on the ingresscontroller to
+		// clearly indicate we're not going to process it.
+		return reconcile.Result{}, nil
+	}
 
-			if err := r.enforceEffectiveIngressDomain(ingress, ingressConfig); err != nil {
-				errs = append(errs, fmt.Errorf("failed to enforce the effective ingress domain for ingresscontroller %s: %v", ingress.Name, err))
-			} else if IsStatusDomainSet(ingress) {
-				if err := r.enforceEffectiveEndpointPublishingStrategy(ingress, infraConfig); err != nil {
-					errs = append(errs, fmt.Errorf("failed to enforce the effective HA configuration for ingresscontroller %s: %v", ingress.Name, err))
-				} else if ingress.DeletionTimestamp != nil {
-					// Handle deletion.
-					if err := r.ensureIngressDeleted(ingress); err != nil {
-						errs = append(errs, fmt.Errorf("failed to ensure ingress deletion: %v", err))
-					}
-				} else if err := r.enforceIngressFinalizer(ingress); err != nil {
-					errs = append(errs, fmt.Errorf("failed to enforce ingress finalizer %s/%s: %v", ingress.Namespace, ingress.Name, err))
-				} else {
-					// Handle everything else.
-					if err := r.ensureIngressController(ingress, dnsConfig, infraConfig); err != nil {
-						errs = append(errs, fmt.Errorf("failed to ensure ingresscontroller: %v", err))
-					}
-				}
-			}
+	// If the ingresscontroller is deleted, handle that and return early.
+	if ingress.DeletionTimestamp != nil {
```

Comment:
TODO: if the ingress controller is invalid, do we still need to delete?

Reply:
Do we need to add a finalizer to invalid objects? If not then invalid objects don't need to go through deletion/finalization